### PR TITLE
listLearners, now with table option

### PR DESCRIPTION
**Pull request by berndbischl**
the new options enhances listLearners with an option to create a table of learners and properties. 
this is all done without loading packages.

it basically allows all props to be accessed that one might want to know about - except for param sets.

### CODE REVIEW

**Comment by berndbischl** [comment]
Can this please be reviewed?
It is also relevant for @schiffner and the tutorial. 

What I still need to do: NEWS + squash my commits

**Comment by larskotthoff** [comment]
Could we simplify the arguments? I'm wondering if there's a case where we don't want to create the learners and don't want a table. You can always extract the names easily from the data frame. So then we could get rid of the `table` argument and say that if `create` is `FALSE`, we return the table data frame.

**Comment by berndbischl** [comment]
I know, but that would break existing API.

**Comment by larskotthoff** [comment]
I see your point, but I'm in favour of this change. It makes the code easier to maintain and I don't think that this particular function is used that often.

**Comment by berndbischl** [comment]
I'll make the change, if you check whether the written part of the tutorial thens needs to be updated.

**Comment by berndbischl** [comment]
I am also not passing the R CMD check, will fix this too

**Comment by larskotthoff** [comment]
Will do.

File: R/listLearners.R [hunk 1]
```diff
@@ -43,13 +48,16 @@
 #' }
 #' @export
 listLearners  = function(obj = NA_character_, properties = character(0L),
-  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE) {
+  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE, table = FALSE) {
 
   if (!missing(obj))
     assert(checkCharacter(obj), checkClass(obj, "Task"))
   assertSubset(properties, getSupportedLearnerProperties())
```
Comment (mllg) [inline R/listLearners.R]: Missing assert for `quiet`.

File: R/listLearners.R [hunk 2]
```diff
@@ -43,13 +48,16 @@
 #' }
 #' @export
 listLearners  = function(obj = NA_character_, properties = character(0L),
-  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE) {
+  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE, table = FALSE) {
 
   if (!missing(obj))
     assert(checkCharacter(obj), checkClass(obj, "Task"))
```
Comment (mllg) [inline R/listLearners.R]: We do a S3 dispatch, so we do not need this assertion.

File: R/listLearners.R (hunk 1 above)
Comment (mllg) [inline R/listLearners.R]: And for `check.packages`

File: R/listLearners.R [hunk 3]
```diff
@@ -57,15 +65,15 @@ listLearners  = function(obj = NA_character_, properties = character(0L),
...
 }
 
 #' @export
 #' @rdname listLearners
 listLearners.character  = function(obj, properties = character(0L),
-  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE) {
+  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE, table = FALSE) {
 
   assertChoice(obj, choices = c("classif", "regr", "surv", "costsens", "cluster", "multilabel", NA_character_))
   assertSubset(properties, getSupportedLearnerProperties(obj))
```
Comment (mllg) [inline R/listLearners.R]: Already checked in dispatching S3.

File: R/listLearners.R [hunk 4]
```diff
@@ -57,15 +65,15 @@ listLearners  = function(obj = NA_character_, properties = character(0L),
...
+  listLearners.character(obj = NA_character_, properties, quiet, warn.missing.packages, check.packages, create, table)
 }
 
 #' @export
 #' @rdname listLearners
 listLearners.character  = function(obj, properties = character(0L),
-  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE) {
+  quiet = TRUE, warn.missing.packages = TRUE, check.packages = TRUE, create = FALSE, table = FALSE) {
 
   assertChoice(obj, choices = c("classif", "regr", "surv", "costsens", "cluster", "multilabel", NA_character_))
```
Comment (mllg) [inline R/listLearners.R]: There should be a global constant / a register for task types.

File: R/listLearners.R [hunk 5]
```diff
@@ -75,6 +83,8 @@ listLearners.character  = function(obj, properties = character(0L),
   # make really sure we filter out our own mock learners here, they have a very unique name
   meths = meths[!grepl("__mlrmocklearners__", meths)]
   res = err = vector("list", length(meths))
+  res.table = data.frame()
+  res.table.props = c(numerics = FALSE, factors = FALSE, ordered = FALSE, missings = FALSE, weights = FALSE, twoclass = FALSE, multiclass = FALSE, class.weights = FALSE, prob = FALSE, se = FALSE, lcens = FALSE, rcens = FALSE, icens = FALSE)
   learner.classes = vcapply(strsplit(meths, "makeRLearner\\."), function(x) x[2L])
   if (!create && check.packages) {
     installed.packs = rownames(installed.packages())
```
Comment (mllg) [inline R/listLearners.R]: We should not call `installed.packages()` (think of large libraries), but use `find.package()` instead.

File: R/listLearners.R [hunk 6]
```diff
@@ -75,6 +83,8 @@ listLearners.character  = function(obj, properties = character(0L),
   # make really sure we filter out our own mock learners here, they have a very unique name
   meths = meths[!grepl("__mlrmocklearners__", meths)]
   res = err = vector("list", length(meths))
+  res.table = data.frame()
+  res.table.props = c(numerics = FALSE, factors = FALSE, ordered = FALSE, missings = FALSE, weights = FALSE, twoclass = FALSE, multiclass = FALSE, class.weights = FALSE, prob = FALSE, se = FALSE, lcens = FALSE, rcens = FALSE, icens = FALSE)
```
Comment (mllg) [inline R/listLearners.R]: This should not be hardcoded.

**Comment by berndbischl** [comment]
> Missing assert for quiet. And for check.packages

Done. 

> assertSubset(properties, getSupportedLearnerProperties(obj)
> Already checked in dispatching S3.

Thats actually not true. The calls are different.

**Comment by berndbischl** [comment]
> There should be a global constant / a register for task types.

I added this in helpers.R: getSupportedTaskTypes
Done.

**Comment by berndbischl** [comment]
> use find.package()

I have done that, pls check extra carefully

**Comment by berndbischl** [comment]
> Regarding the props in the result table
> This should not be hardcoded.

Ok. We aready had getSupportedLearnerProperties and where using this at another place in the code.
So I am using it here. 
Please note: We also have to maintain such stuff. 
You already forgot to put some of your surv properties in there. I am therefore also kinda sure that you are missing some unit tests for this.
Please check.
Otherwise: done.

**Comment by berndbischl** [comment]
@mllg @larskotthoff 
Thx for the detailled review. I should be done now.

Any further comments?

**Comment by larskotthoff** [comment]
The API change should be documented in NEWS, but I'm happy to add that after merging this. Fine with me apart from that.

**Comment by berndbischl** [comment]
NEWS and squash I will do, of course. I am waiting for Michel. But he is also sick at the moment. 
But I guess it is not urgent to merge this in 24h now.

**Comment by larskotthoff** [comment]
Please don't merge unfinished PRs with failing tests!

**Comment by mllg** [comment]
Sorry for that. I've checked travis and it looked okay (<https://travis-ci.org/mlr-org/mlr/builds/96885131>: only a warning that the Rd is not updated which travis should update automatically after the merge, right?). Maybe Travis did not start a new build after the last commits? Don't know, will be more careful in the future.

**Comment by berndbischl** [comment]
It was fifty fifty our fault, Michel said in review everything looked ok, then I asked him to merge.

But please: Read through the thread. It is obvious that I still did not squash. And NEWS is not updated. This is unfortunately the reviewer responsibility. Now we have 3 commits with ... as messages. 

I now might have to rebase master what I hate

**Comment by larskotthoff** [comment]
I have already updated NEWS and fixed some of the failing tests -- I should be able to fix the rest by tonight. I'm not in favour of messing with the history on master.

**Comment by larskotthoff** [comment]
Failures are fixed now.

**Comment by schiffner** [comment]
Thank you all. I tried it out with the tutorial and that works well.

The "old" tables in the tutorial also show the `note`, which is not returned now by `listLearners`. I don't have a strong opinion about this, but including it might be useful and is easily done.
I noticed some small things in the docs (grammar, typos). Also multilabel is not mentioned.
If you agree I can do a PR.

**Comment by larskotthoff** [comment]
Sounds good, please go ahead.

**Comment by schiffner** [comment]
Will do tomorrow. Will go to sleep now. :)

**Comment by berndbischl** [comment]
thank you. sorry for forgetting "note", it should be in there.
and fixing my bad grammar and spelling would be much appreciated :)

**Comment by schiffner** [comment]
:wink:

**Comment by berndbischl** [comment]
> Im not in favour of messing with the history on master.

I guess we can live with the bad messages, and I agree that I dislike this too.

Is there actually an easy "undo" for merges (assuming it was the last action on the master branch)?

**Comment by larskotthoff** [comment]
There's a "revert" button that probably does just that.

**Comment by berndbischl** [comment]
hmm, maybe we should look into this later, so we can repair such stuff

also i am really not sure why i am posting about this at 2am in the night

**Pull request merged**